crates/ruma-federation-api/src/authenticated_media.rs

```diff
@@ -0,0 +1,291 @@
+    http_response: http::Response<T>,
+) -> Result<
+    (ContentMetadata, FileOrLocation),
+    ruma_common::api::error::FromHttpResponseError<ruma_common::api::error::MatrixError>,
+> {
+    use mail_parser::{parsers::MessageStream, GetHeader, HeaderName, HeaderValue, MessageParser};
+    use ruma_common::api::error::{HeaderDeserializationError, MultipartMixedDeserializationError};
+
+    let mut content_type_value = http_response
+        .headers()
+        .get(http::header::CONTENT_TYPE)
+        .ok_or_else(|| HeaderDeserializationError::MissingHeader("Content-Type".to_owned()))?
+        .to_str()?
+        .to_owned();
+    // MessageStream expects the header to end with a newline.
+    if !content_type_value.ends_with('\n') {
+        content_type_value.push('\n');
+    }
+
+    let mut content_type_stream = MessageStream::new(content_type_value.as_bytes());
+    let content_type = match content_type_stream.parse_content_type() {
+        HeaderValue::ContentType(content_type)
+            if content_type.ctype() == "multipart" && content_type.subtype() == Some("mixed") =>
+        {
+            content_type
+        }
+        _ => {
+            return Err(HeaderDeserializationError::InvalidHeaderValue {
+                header: "Content-Type".to_owned(),
+                expected: "multipart/mixed".to_owned(),
+                unexpected: content_type_value.to_owned(),
+            }
+            .into())
+        }
+    };
+
+    let boundary = content_type
+        .attribute("boundary")
+        .ok_or(HeaderDeserializationError::MissingMultipartBoundary)?
+        .as_bytes();
+
+    let mut stream = MessageStream::new(http_response.body().as_ref());
+    // The parser is only used for the configuration to parse headers.
+    let parser = MessageParser::new()
+        .header_text(HeaderName::ContentType)
+        .header_text(HeaderName::ContentDisposition)
+        .header_text("Location");
+
+    // Advance to first part.
+    if !stream.seek_next_part(boundary) {
+        return Err(MultipartMixedDeserializationError::MissingBodyParts {
+            expected: 2,
+            unexpected: 0,
+        }
+        .into());
+    }
+    stream.skip_crlf();
+
+    // The first part should be the metadata serialized as JSON.
+    let mut metadata_headers = Vec::new();
+    if !stream.parse_headers(&parser, &mut metadata_headers)
+        || metadata_headers.header_value(&HeaderName::ContentType).is_none()
+    {
+        return Err(
+            MultipartMixedDeserializationError::MissingHeader("Content-Type".to_owned()).into()
+        );
+    }
+
+    let (_, metadata_raw) = stream.mime_part(boundary);
+    let metadata = serde_json::from_slice::<ContentMetadata>(&metadata_raw)?;
+
+    stream.skip_crlf();
+
+    // The second part should be the content or a redirect.
+    let mut content_headers = Vec::new();
+    let location_header_name =
+        HeaderName::parse("location").expect("should be a valid header name");
```

Comment:
Considering we can just do this, it makes me wonder why the `parse` function even exists.
```suggestion
    let location_header_name = HeaderName::from("Location");
```

Reply:
I don't like this `From` implementation, it doesn't actually make sense. It basically does `HeaderName::parse(string).unwrap_or_default()`, which means that an error in the header name could actually go unseen until you actually test the code in production. That's why I prefer to `.expect` here.

Reply:
Ah ok, that makes sense. Does the capitalization matter here? It seems like `This-Case` is used as standard.

Reply:
The header name should be case-insensitive. I don't really have a preference here.

Reply:
Not used anymore.